### DISCUSSION TITLE
Jib plugin version upgrade

### DISCUSSION THREAD

**Discussion by skumaravelan** [apache/camel-karavan#997] [Q&A] [answered]
How to update Jib plugin, current on uses [3.3.2], I want  to update 3.4.0.

**Comment by davsclaus** [comment]
This will be possible from Camel 4.2 onwards
https://issues.apache.org/jira/browse/CAMEL-20026

See the TIP at
https://camel.apache.org/manual/camel-jbang.html#_exporting_to_camel_main_for_kubernetes

The next version of Karavan is planned to upgrade to Camel 4.2

**Reply by skumaravelan** [reply]
@davsclaus I want to add the below snippet during the jbang export. what are the steps to add these entries in the jib plugin configuration section. 
`     <from>
                        <image>${jib.from.image}</image>
                    </from>
                    <to>
                        <image>${jib.to.image}</image>
                        <auth>
                            <username>${jib.to.auth.username}</username>
                            <password>${jib.to.auth.password}</password>
                        </auth>
                    </to>`

**Comment by skumaravelan** [comment]
@davsclaus I have the following values in my application properties, am still not seeing a change in my generated pom.xml. 
camel.jbang.version= 4.2.0
camel.jbang.jib-maven-plugin-version=3.4.0

**Reply by skumaravelan** [reply]
In addition to that I set CAMEL_VERSION=4.2.0 and i see the plugin version updated in the pom. Thank you for the pointer.

**Comment by davsclaus** [comment]
Ah yeah if you upgraded to jbang camel 4.2.0 as its the "thing" that does the export and creates the pom.xml so then you can control it now

**Reply by skumaravelan** [reply]
But that is not working.. What could be an issue.

**Comment by skumaravelan** [comment]
Got it working now by adding CAMEL_VERSION=4.2.0  in build.sh .

**Comment by skumaravelan** [comment]
@davsclaus :  am looking for a way to add this section in jib plugin configuration. application properties is adding them into to config section.
```
     <from>
                        <image>${jib.from.image}</image>
                    </from>
                    <to>
                        <image>${jib.to.image}</image>
                        <auth>
                            <username>${jib.to.auth.username}</username>
                            <password>${jib.to.auth.password}</password>
                        </auth>
                    </to>
```

**Reply by skumaravelan** [reply]
When I use this maven command it is failing with No plugin found for prefix 'alfresco' in the current project and in the plugin groups [org.apache] with the default pom export template. Now I need a mechanism to add the  above elements. Please suggest. 

mvn -U -X -s "${BUILD_SECRETS_PATH}/settings.xml" package jib:build
          -Djib.allowInsecureRegistries=true \
          -Djib.httpTimeout=180000 \
          -Djib.to.auth.username=${IMAGE_REGISTRY_USERNAME} \
          -Djib.to.auth.password=${IMAGE_REGISTRY_PASSWORD}

**Comment by skumaravelan** [comment]
I am blocked by this issue. Pleases share  Any workaround suggestions to get pass this issue.

**Comment by skumaravelan** [comment]
Any way to pass the template as a configuration or something?

**Comment by davsclaus** [comment]
Its not possible currently. As the build is just a script file (sh) you can edit this to manipulate the files after the export.

To add support for more flexibile exports then this needs to be added as a ticket and implemented in camel-jbang

**Comment by davsclaus** [comment]
Can you paste below the full XML output what you want it to be liked that can help us to ensure we implement something that can support this use-case

**Comment by skumaravelan** [comment]
`<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>org.camel.karavan.demo</groupId>
    <artifactId>apple-hello</artifactId>
    <version>1</version>

    <properties>
        <java.version>17</java.version>
        <label.runtime>app.kubernetes.io/runtime</label.runtime>
        <jib.from.image>docker.apple.com/base-images/ubi9/java17-builder:latest</jib.from.image>
        <jkube.version>1.14.0</jkube.version>
        <jkube.skip.build>true</jkube.skip.build>
        <jkube.namespace>default</jkube.namespace>
        <jkube.imagePullPolicy>IfNotPresent</jkube.imagePullPolicy>
        <jkube.enricher.jkube-controller.type>Deployment</jkube.enricher.jkube-controller.type>
        <jkube.enricher.jkube-controller.replicaCount>1</jkube.enricher.jkube-controller.replicaCount>
        <jkube.enricher.jkube-service.port>80:8080</jkube.enricher.jkube-service.port>
        <jkube.enricher.jkube-project-label.group>karavan</jkube.enricher.jkube-project-label.group>
        <jkube.enricher.jkube-project-label.provider>karavan</jkube.enricher.jkube-project-label.provider>
        <jkube.recreate>true</jkube.recreate>
        <jkube.serviceAccountName>default</jkube.serviceAccountName>
        <jib.to.image>docker.apple.com/stratos-workflow/apple-hello</jib.to.image>

    </properties>

    <dependencyManagement>
        <dependencies>
            
            <dependency>
                <groupId>org.apache.camel</groupId>
                <artifactId>camel-bom</artifactId>
                <version>4.2.0</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>



    <dependencies>

        <dependency>
            <groupId>org.apache.camel</groupId>
            <artifactId>camel-main</artifactId>
        </dependency>
        <dependency>
            <groupId>org.apache.camel</groupId>
            <artifactId>camel-health</artifactId>
        </dependency>
        <dependency>
            <groupId>org.apache.camel</groupId>
            <artifactId>camel-console</artifactId>
        </dependency>
        <dependency>
            <groupId>org.apache.camel</groupId>
            <artifactId>camel-platform-http-main</artifactId>
        </dependency>
        <dependency>
            <groupId>org.apache.camel</groupId>
            <artifactId>camel-platform-http-main</artifactId>
        </dependency>
        <dependency>
            <groupId>org.apache.camel</groupId>
            <artifactId>camel-rest</artifactId>
        </dependency>
        <dependency>
            <groupId>org.apache.camel</groupId>
            <artifactId>camel-yaml-dsl</artifactId>
        </dependency>


        
        <dependency>
            <groupId>org.fusesource.jansi</groupId>
            <artifactId>jansi</artifactId>
            <version>2.4.0</version>
        </dependency>
        <dependency>
            <groupId>org.apache.logging.log4j</groupId>
            <artifactId>log4j-api</artifactId>
            <version>2.20.0</version>
        </dependency>
        <dependency>
            <groupId>org.apache.logging.log4j</groupId>
            <artifactId>log4j-slf4j2-impl</artifactId>
            <version>2.20.0</version>
        </dependency>
        <dependency>
            <groupId>org.apache.logging.log4j</groupId>
            <artifactId>log4j-core</artifactId>
            <version>2.20.0</version>
        </dependency>

        <dependency>
            <groupId>org.apache.camel</groupId>
            <artifactId>camel-test-junit5</artifactId>
            <scope>test</scope>
        </dependency>
    </dependencies>

    <build>
        <plugins>

            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.11.0</version>
                <configuration>
                    <source>${java.version}</source>
                    <target>${java.version}</target>
                </configuration>
            </plugin>
            
            <plugin>
                <groupId>org.apache.camel</groupId>
                <artifactId>camel-maven-plugin</artifactId>
                <version>4.2.0</version>
                <configuration>
                    <mainClass>org.camel.karavan.demo.applehello.CamelApplication</mainClass>
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>prepare-fatjar</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-assembly-plugin</artifactId>
                <version>3.6.0</version>
                <configuration>
                    <descriptors>
                        <descriptor>src/main/resources/assembly/runner.xml</descriptor>
                    </descriptors>
                    <archive>
                        <manifest>
                            <mainClass>org.camel.karavan.demo.applehello.CamelApplication</mainClass>
                        </manifest>
                        <manifestEntries>
                            <Multi-Release>true</Multi-Release>
                        </manifestEntries>
                    </archive>
                    <appendAssemblyId>false</appendAssemblyId>
                </configuration>
                <executions>
                    <execution>
                        <id>make-assembly</id>
                        <phase>package</phase>
                        <goals>
                            <goal>single</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            
            <plugin>
                <groupId>com.google.cloud.tools</groupId>
                <artifactId>jib-maven-plugin</artifactId>
                <version>3.4.0</version>
                <configuration>
                    <from>
                        <image>${jib.from.image}</image>
                    </from>
                    <to>
                        <image>${jib.to.image}</image>
                        <auth>
                            <username>${jib.to.auth.username}</username>
                            <password>${jib.to.auth.password}</password>
                        </auth>
                    </to>
                    <skip>false</skip>
                    <containerizingMode>packaged</containerizingMode>
                    <container>
                        <ports>
                            <port>8080</port>
                        </ports>
                    </container>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.eclipse.jkube</groupId>
                <artifactId>kubernetes-maven-plugin</artifactId>
                <version>${jkube.version}</version>
                <configuration>
                    <images>
                        <image>
                            <name>${jib.to.image}</name>
                            <build>
                            </build>
                        </image>
                    </images>
                    <resources>
                        <labels>
                            <all>
                                <property>
                                    <name>${label.runtime}</name>
                                    <value>camel</value>
                                </property>
                            </all>
                        </labels>
                    </resources>
                </configuration>
            </plugin>

        </plugins>
    </build>

    <profiles>
        <profile>
            <id>camel.debug</id>
            <activation>
                <property>
                    <name>camel.debug</name>
                    <value>true</value>
                </property>
            </activation>
            <dependencies>
                <dependency>
                    <groupId>org.apache.camel</groupId>
                    <artifactId>camel-debug</artifactId>
                </dependency>
            </dependencies>
        </profile>
    </profiles>

</project>
`

**Comment by skumaravelan** [comment]
This is the part needed in addition to generated one. 

```
                <from>
                    <image>${jib.from.image}</image>
                </from>
                <to>
                    <image>${jib.to.image}</image>
                    <auth>
                        <username>${jib.to.auth.username}</username>
                        <password>${jib.to.auth.password}</password>
                    </auth>
                </to>
```

**Comment by davsclaus** [comment]
I have created a ticket
https://issues.apache.org/jira/browse/CAMEL-20125

**Reply by skumaravelan** [reply]
Is it ready to test and use now? Let me know, how can we use this version and add these new configuration? 
how to pass these config as part of the export command?

**Reply by davsclaus** [reply]
You would need to use Camel version 4.3.0-SNAPSHOT
And you can set these options in the application.properties file

The SNAPSHOT release is downloaded from: http://repository.apache.org/content/groups/snapshots/
Which is slower and is daily updated but if there is some hickups then it may not be fully up to date. And the fix was committed yesterday. However last build was green and includes this fix
https://ci-builds.apache.org/job/Camel/job/Camel%20Daily%20Snapshot%20Deploy/job/main/lastBuild/

**Reply by skumaravelan** [reply]
When is the official release of the this fix and when would be available as 4.3.0.?

**Reply by davsclaus** [reply]
Camel 4.3.0 is scheduled for next month, and 4.4.0 LTS in Jan/Feb 2024.

**Reply by skumaravelan** [reply, marked as answer]
Thank you. That would work for me.
